File: Telegram.fs
module comicdownloader.Telegram

open System.Threading.Tasks
open Telegram.Bot
open Telegram.Bot.Types
open System.Net.Http
open System.IO
open Microsoft.Extensions.Logging

let client = new HttpClient()

type TelegramSettings = {
    ApiKey: string
    User: string
}

let sendText (logger: ILogger) settings (message: string) =
    match settings with
    | None ->
        logger.LogWarning("Skipping Telegram text message because no Telegram settings were found")
        Task.FromResult(Error (exn("No settings")))
    | Some settings ->
        let telegramClient = TelegramBotClient(settings.ApiKey, client)
        task {
            try
                logger.LogInformation("Sending Telegram text message")
                let! _ = telegramClient.SendMessage(settings.User, message)
                logger.LogInformation("Telegram text message sent")
                return Ok "Sent"
            with
                | exn ->
                    logger.LogError(exn, "Telegram text message failed")
                    return Error exn
        }

let sendMessage (logger: ILogger) settings (image: byte array) =
    match settings with
    | None ->
        logger.LogWarning("Skipping Telegram image send because no Telegram settings were found")
        Task.FromResult(Error (exn("No settings")))
    | Some settings ->
        let telegramClient = TelegramBotClient(settings.ApiKey, client)
        task {
            try
                use ms = new MemoryStream(image)
                logger.LogInformation("Sending Telegram image with {ByteCount} bytes", image.Length)
                let! _ = telegramClient.SendPhoto(settings.User, InputFile.FromStream(ms))
                logger.LogInformation("Telegram image sent")
                return Ok "Sent"
            with
                | exn ->
                    logger.LogError(exn, "Telegram image send failed")
                    return Error exn
        }

File: telegram.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
)

type TelegramClient struct {
	bot    *tgbotapi.BotAPI
	chatID int64
}

func NewTelegramClient(settings *TelegramSettings) (*TelegramClient, error) {
	chatID, err := strconv.ParseInt(settings.User, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid Telegram User %q: %w", settings.User, err)
	}
	bot, err := tgbotapi.NewBotAPI(settings.ApiKey)
	if err != nil {
		return nil, fmt.Errorf("create telegram bot: %w", err)
	}
	return &TelegramClient{bot: bot, chatID: chatID}, nil
}

func normalizeCaption(text string) *string {
	if strings.TrimSpace(text) == "" {
		return nil
	}
	if len(text) > 1024 {
		text = text[:1024]
	}
	return &text
}

func (t *TelegramClient) SendPhoto(data []byte, filename string, caption *string) error {
	msg := tgbotapi.NewPhoto(t.chatID, tgbotapi.FileBytes{Name: filename, Bytes: data})
	if caption != nil {
		msg.Caption = *caption
	}
	_, err := t.bot.Send(msg)
	return err
}

func (t *TelegramClient) SendText(text string) error {
	if len(text) > 4000 {
		text = text[:4000]
	}
	_, err := t.bot.Send(tgbotapi.NewMessage(t.chatID, text))
	return err
}
